packages/api-contract/src/MetaRegistry.ts

```diff
@@ -0,0 +1,333 @@
+
+    return this._strings[index - 1];
+  }
+
+  public stringsAt (indices: t.StringIndex[]): string[] {
+    return indices.map(
+      (index: number): string => {
+        return this.stringAt(index);
+      }
+    );
+  }
+
+  public typeAt (index: t.TypeIndex): t.MetaType {
+    assert(this._types && index <= this._types.length, `MetaRegistry: Invalid type index ${index} found in metadata`);
+
+    return this._types[index - 1];
+  }
+
+  public typesAt (indices: t.TypeIndex[]): t.MetaType[] {
+    return indices.map(
+      (index: number): t.MetaType => {
+        return this.typeAt(index);
+      }
+    );
+  }
+
+  public hasTypeDefAt (index: t.TypeIndex): boolean {
+    return !!this.typeDefs[index - 1];
+  }
+
+  public typeDefAt (index: t.TypeIndex, extra: Pick<t.TypeDef, never> = {}): t.TypeDef {
+    assert(this.hasTypeDefAt(index), `Invalid TypeDef index requested: ${index}`);
+    return {
+      ...this.typeDefs[index - 1],
+      ...extra
+    };
+  }
+
+  private typeDefIdFields ({ id }: t.MetaType): Pick<t.TypeDef, never> {
+    if (!id) {
+      return {};
+    }
+
+    const {
+      'custom.name': nameIndex,
+      'custom.namespace': namespaceIndices,
+      'custom.params': paramsIndices
+    } = id as t.MetaTypeIdCustom;
+
+    if (nameIndex) {
+      return {
+        name: this.stringAt(nameIndex),
+        ...(
+          namespaceIndices && namespaceIndices.length
+            ? { namespace: this.stringsAt(namespaceIndices).join('::') }
+            : {}
+        ),
+        ...(
+          paramsIndices && paramsIndices.length
+            ? { params: this.typesAt(paramsIndices).map((type): t.TypeDef => this.typeDefFromMetaType(type)) }
+            : {}
+        ),
+        ...(SPECIAL_TYPES.includes(this.stringAt(nameIndex))
+          ? { type: this.stringAt(nameIndex) }
+          : {}
+        )
+      };
+    }
+
+    return {};
+  }
+
+  private typeDefDefFields (metaType: t.MetaType, typeIndex?: t.TypeIndex): Pick<t.TypeDef, never> {
+    assert(!(metaType.def as t.MetaTypeDefUnion)['union.fields'], 'Invalid union type definition found');
+
+    let typeDef: Pick<t.TypeDef, never> = {};
+    if (this.isBuiltin(metaType)) {
+      typeDef = this.typeDefForBuiltin(metaType, typeIndex);
+    }
+
+    if (this.isEnum(metaType)) {
+      typeDef = this.typeDefForEnum(metaType.def as t.MetaTypeDefEnum);
+    }
+
+    if (this.isClikeEnum(metaType)) {
+      typeDef = this.typeDefForClikeEnum(metaType.def as t.MetaTypeDefClikeEnum);
+    }
+
+    if (this.isStruct(metaType)) {
+      typeDef = this.typeDefForStruct(metaType.def as t.MetaTypeDefStruct);
+    }
+
+    if (this.isTupleStruct(metaType)) {
+      typeDef = this.typeDefForTupleStruct(metaType.def as t.MetaTypeDefTupleStruct);
+    }
+
+    assert(Object.keys(typeDef).length > 0, 'Error decoding metadata type');
+    return typeDef || {};
```

Comment:
As per the comment above, when we get here, it should be defined. We could generally actually just get rid of the assinments here and the just 

```js
  else if (this.isTupleStruct(metaType)) {
      return this.typeDefForTupleStruct(metaType.def as t.MetaTypeDefTupleStruct);
    }
```